app/gui2/src/components/visualizations/HistogramVisualization.vue

```diff
@@ -0,0 +1,675 @@
+  x0: number | undefined
+  x1: number | undefined
+  length: number | undefined
+}
+</script>
+
+<script setup lang="ts">
+import { computed, onMounted, ref, watch, watchEffect, watchPostEffect } from 'vue'
+
+// @ts-expect-error
+// eslint-disable-next-line no-redeclare
+import * as d3 from 'https://cdn.jsdelivr.net/npm/d3@7.8.5/+esm'
+
+import type {
+  BrushSelection,
+  D3BrushEvent,
+  D3ZoomEvent,
+  ScaleContinuousNumeric,
+  ScaleSequential,
+  ZoomTransform,
+} from 'd3'
+
+import SvgIcon from '@/components/SvgIcon.vue'
+import VisualizationContainer from '@/components/VisualizationContainer.vue'
+import { useVisualizationConfig } from '@/providers/useVisualizationConfig.ts'
+
+import { updateableRef } from './builtins.ts'
+import { useEvent, useEventConditional } from './events.ts'
+import { getTextWidth } from './measurement.ts'
+
+const shortcuts = {
+  zoomIn: (e: KeyboardEvent) => (e.ctrlKey || e.metaKey) && e.key === 'z',
+  showAll: (e: KeyboardEvent) => (e.ctrlKey || e.metaKey) && e.key === 'a',
+}
+
+const MARGIN = 25
+const AXIS_LABEL_HEIGHT = 10
+const ANIMATION_DURATION_MS = 400
+const DEFAULT_NUMBER_OF_BINS = 50
+const COLOR_LEGEND_WIDTH = 5
+const DEFAULT_AXES_CONFIGURATION: AxesConfiguration = {
+  x: { scale: ScaleType.Linear },
+  y: { scale: ScaleType.Linear },
+}
+const RMB_DIVIDER = 100
+const PINCH_DIVIDER = 100
+
+const EPSILON = 0.001
+const ZOOM_EXTENT = [0.5, 20] satisfies BrushSelection
+const RIGHT_BUTTON = 2
+const MID_BUTTON = 1
+const MID_BUTTON_CLICKED = 4
+const SCROLL_WHEEL = 0
+
+const props = defineProps<{ data: Data }>()
+const emit = defineEmits<{
+  'update:preprocessor': [module: string, method: string, ...args: string[]]
+}>()
+
+const config = useVisualizationConfig()
+
+const xAxisNode = ref<SVGGElement>()
+const yAxisNode = ref<SVGGElement>()
+const plotNode = ref<SVGGElement>()
+const colorLegendGradientNode = ref<SVGElement>()
+const zoomNode = ref<SVGGElement>()
+const brushNode = ref<SVGGElement>()
+
+const d3XAxis = computed(() => d3.select(xAxisNode.value))
+const d3YAxis = computed(() => d3.select(yAxisNode.value))
+const d3Plot = computed(() => d3.select(plotNode.value))
+const d3ColorLegendGradient = computed(() => d3.select(colorLegendGradientNode.value))
+const d3Zoom = computed(() => d3.select(zoomNode.value))
+const d3Brush = computed(() => d3.select(brushNode.value))
+
+const points = ref<number[]>([])
+const rawBins = ref<number[]>()
+const binCount = ref(DEFAULT_NUMBER_OF_BINS)
+const axis = ref(DEFAULT_AXES_CONFIGURATION)
+const rawFocus = ref<Focus>()
+const brushExtent = ref<BrushSelection>()
+const zoomLevel = ref(1)
+const shouldAnimate = ref(false)
+
+const originalXScale = updateableRef(d3.scaleLinear())
+const xScale = updateableRef(d3.scaleLinear())
+const yScale = updateableRef(d3.scaleLinear())
+const fill = updateableRef(d3.scaleSequential().interpolator(d3.interpolateViridis))
+const yAxis = updateableRef(d3.axisLeft(yScale.value).tickFormat(d3.format('d')))
```

Comment:
Since scales and axes are really stateful closures, we can use `computed` instead of relying on their internal state changes. That way we can keep reactivity working as expected.

```suggestion
const originalXScale = computed(() => {
    return d3.scaleLinear().range([0, boxWidth.value]).domain(xExtents.value)
})
const xScale = computed(() => {
    d3.scaleLinear().domain(xDomain.value).range([0, boxWidth.value])
})
const yScale = computed(() => {
    d3.scaleLinear().domain(yDomain.value).range([boxHeight.value, 0])
})
const fill = computed(() => {
    d3.scaleSequential().domain(yDomain.value).interpolator(d3.interpolateViridis)
})
const yAxis = computed(() => {
    const yTicks = yScale.value.ticks().filter(Number.isInteger)
    d3.axisLeft(yScale.value).tickFormat(d3.format('d')).tickValues(yTicks)
})
````

Note that if this approach causes any further issues, I'm fine with current workaround too (just use normal `ref`s and `updateRef` instead).